pkg/client/actionclient.go

```diff
@@ -61,13 +61,71 @@ type GetOption func(*action.Get) error
 type InstallOption func(*action.Install) error
 type UpgradeOption func(*action.Upgrade) error
 type UninstallOption func(*action.Uninstall) error
+type RollbackOption func(*action.Rollback) error
 
-func NewActionClientGetter(acg ActionConfigGetter) ActionClientGetter {
-	return &actionClientGetter{acg}
+type ActionClientGetterOption func(*actionClientGetter) error
+
+func AppendGetOptions(opts ...GetOption) ActionClientGetterOption {
+	return func(getter *actionClientGetter) error {
+		getter.defaultGetOpts = append(getter.defaultGetOpts, opts...)
+		return nil
+	}
+}
+
+func AppendInstallOptions(opts ...InstallOption) ActionClientGetterOption {
+	return func(getter *actionClientGetter) error {
+		getter.defaultInstallOpts = append(getter.defaultInstallOpts, opts...)
+		return nil
+	}
+}
+
+func AppendUpgradeOptions(opts ...UpgradeOption) ActionClientGetterOption {
+	return func(getter *actionClientGetter) error {
+		getter.defaultUpgradeOpts = append(getter.defaultUpgradeOpts, opts...)
+		return nil
+	}
+}
+
+func AppendUninstallOptions(opts ...UninstallOption) ActionClientGetterOption {
+	return func(getter *actionClientGetter) error {
+		getter.defaultUninstallOpts = append(getter.defaultUninstallOpts, opts...)
+		return nil
+	}
+}
+
+func AppendInstallFailureUninstallOptions(opts ...UninstallOption) ActionClientGetterOption {
+	return func(getter *actionClientGetter) error {
+		getter.installFailureUninstallOpts = append(getter.installFailureUninstallOpts, opts...)
+		return nil
+	}
+}
+func AppendUpgradeFailureRollbackOptions(opts ...RollbackOption) ActionClientGetterOption {
+	return func(getter *actionClientGetter) error {
+		getter.upgradeFailureRollbackOpts = append(getter.upgradeFailureRollbackOpts, opts...)
+		return nil
+	}
+}
+
+func NewActionClientGetter(acg ActionConfigGetter, opts ...ActionClientGetterOption) (ActionClientGetter, error) {
+	actionClientGetter := &actionClientGetter{acg: acg}
+	for _, opt := range opts {
+		if err := opt(actionClientGetter); err != nil {
+			return nil, err
+		}
+	}
+	return actionClientGetter, nil
 }
 
 type actionClientGetter struct {
 	acg ActionConfigGetter
+
+	defaultGetOpts       []GetOption
+	defaultInstallOpts   []InstallOption
+	defaultUpgradeOpts   []UpgradeOption
+	defaultUninstallOpts []UninstallOption
+
+	installFailureUninstallOpts []UninstallOption
+	upgradeFailureRollbackOpts  []RollbackOption
```

Comment:
Should we initialize these newly added slice fields in the `actionClientGetter` returned from `NewActionClientGetter()`? 

I'm not familiar with how this `ActionClientGetter` is typically used, but I am wondering - is there any cases that could occur where a user is attempting to do something with one of these new slice fields but the program panics because the slices haven't been initialized (IIRC in Go, slices uninitialized default value is `nil`)?

Reply:
> Go, slices uninitialized default value is nil

This is true, but you can perform slice operations on a `nil` slice with the exact same behavior as if the slice is initialized and has length and capacity 0.

As far as I know, the only panic opportunity is if you try to index a `nil` slice, but that same opportunity exists on an initialized slice with length/capacity 0.